### PR TITLE
Initial block for .NET in-proc indexing in the OOP host.

### LINKED ISSUE #10481: Remove in-proc support from OOP host - Part 1
Item tracking blocking of in-proc indexing from the OOP host (part 1).

Part 2 will remove the capability.

### PR DESCRIPTION
**Pull request by fabiocav**
Immediate fix to ensure proper handling of in-proc payloads in environments where it isn't supported. This defaults to Flex consumption today, with support for hosting configuration.

This currently ignores restrictions for placeholders to avoid potential impact to cold start. The downside is that it misses the opportunity to reduce memory consumption in those cases, but that enhancement will came in #10504 

### Issue describing the changes in this PR

resolves #10481 

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 1]
```diff
@@ -115,6 +115,12 @@ public static bool AzureFilesAppSettingsExist(this IEnvironment environment)
                    !string.IsNullOrEmpty(environment.GetEnvironmentVariable(AzureFilesContentShare));
         }
 
+        public static bool IsDotNetInProcSupported(this IEnvironment environment)
+        {
+            return !IsFlexConsumptionSku(environment)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs:120]: add unit tests?

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 2]
```diff
@@ -115,6 +115,12 @@ public static bool AzureFilesAppSettingsExist(this IEnvironment environment)
...
+                   && !string.Equals(environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsDisableInProc), "1", StringComparison.OrdinalIgnoreCase);
+        }
```
Comment (pragnagopa) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs:122]: Does `EnvironmentSettingNames.FunctionsDisableInProc` default to true for FlexConsumption or does it have to be set to true in dockerfile for FlexConsumption image?

Comment (fabiocav) [reply src/WebJobs.Script/Environment/EnvironmentExtensions.cs:122]: No. This gives you the ability to set this outside of Flex or an environment where a hosting config is applied. Flex will be automatically picked up by the SKU check.

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs (hunk 1 above)
Comment (fabiocav) [reply src/WebJobs.Script/Environment/EnvironmentExtensions.cs:120]: Validation was performed on this. I don't expect we'd have this logic here for longer than a release, but I can follow up with some additional tests.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 3]
```diff
@@ -823,15 +823,33 @@ internal async Task<Collection<FunctionDescriptor>> GetFunctionDescriptorsAsync(
...
+                var inProcIndexingSupported = _environment.IsPlaceholderModeEnabled()
+                                              || (_environment.IsDotNetInProcSupported()
+                                                  && !_hostingConfigOptions.Value.IsDotNetInProcDisabled);
+
                 foreach (FunctionMetadata metadata in functions)
                 {
                     try
                     {
-                        bool created = false;
+                        // If this is metadata represents a function that supports direct type indexing (in-proc), log a warning and skip it.
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: If I'm reading the below code correctly, it looks like we won't skip the function like you say with a warning, but you throw an uncaught error?

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs (hunk 1 above)
Comment (mathewc) [reply src/WebJobs.Script/Environment/EnvironmentExtensions.cs:120]: Yeah, would be good to add tests for these new helpers.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 3 above)
Comment (fabiocav) [reply src/WebJobs.Script/Host/ScriptHost.cs]: That's correct. The comment is stale from an initial version that was going to emit a warning. Will update.

**Review by pragnagopa — APPROVED** [review]

**Review by mathewc — APPROVED** [review]

**Review by pragnagopa — APPROVED** [review]

**Comment by fabiocav** [comment]
@microsoft-github-policy-service rerun

**Pull request merged**